src/Type/DrupalClassResolverReturnType.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace mglaman\PHPStanDrupal\Type;
+
+use mglaman\PHPStanDrupal\Drupal\DrupalServiceDefinition;
+use mglaman\PHPStanDrupal\Drupal\ServiceMap;
+use PhpParser\Node\Expr\CallLike;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+
+final class DrupalClassResolverReturnType
+{
+
+    public static function getType(
+        MethodReflection $methodReflection,
+        CallLike $methodCall,
+        Scope $scope,
+        ServiceMap $serviceMap
+    ): Type {
+        $arg1 = $scope->getType($methodCall->getArgs()[0]->value);
+        if (!$arg1 instanceof ConstantStringType) {
+            return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+        }
+
+        $serviceDefinition = $serviceMap->getService($arg1->getValue());
+        if ($serviceDefinition instanceof DrupalServiceDefinition) {
+            return self::getTypeFromService($serviceDefinition, $arg1->getValue());
+        }
+
+        return new ObjectType($arg1->getValue());
+    }
+
+    private static function getTypeFromService(DrupalServiceDefinition $serviceDefinition, string $serviceId): Type
+    {
+        // Work around Drupal misusing the SplString class for string
+        // pseudo-services such as 'app.root'.
+        // @see https://www.drupal.org/project/drupal/issues/3074585
+        if ($serviceDefinition->getClass() === 'SplString') {
+            return new StringType();
+        }
```

Comment:
Actually, this could be a static method on DrupalServicedefinition, maybe? 

We could provide a `getType` on the definition class. It breaks the "copy and paste" from Drupal core. But whatever, it's forked code anyways. Then we do not need workarounds as the workaround logic is now encapsulated in the class itself.

Reply:
Nice idea to add it to DrupalServiceDefinition!

I choose a non-static method since right now everywhere it is needed we have the DrupalServiceDefinition directly.